### PR TITLE
Ignore sorbet: URIs in textDocument open/change/close events.

### PR DESCRIPTION
**Pull request by jvilk-stripe**
Ignore sorbet: URIs in textDocument open/change/close events. `sorbet:` URIs are sent to VS Code for read-only documents (e.g. autogenerated files or rbis in payload).

### Motivation

Fixes crash. Previously, opening and closing an RBI file would cause Sorbet to commit an empty file with the `sorbet:` URI, and cause all sorts of havoc.

cc @evan-stripe who reported this bug.

### Test plan

See included automated tests.

### CODE REVIEW

File: main/lsp/LSPConfiguration.cc [hunk 1]
```diff
@@ -216,10 +218,13 @@ bool LSPConfiguration::isFileIgnored(string_view filePath) const {
     return FileOps::isFileIgnored(rootPath, filePath, opts.absoluteIgnorePatterns, opts.relativeIgnorePatterns);
 }
 
+bool LSPConfiguration::isSorbetUri(string_view uri) const {
```
Comment (jvilk-stripe) [inline main/lsp/LSPConfiguration.cc]: Exposed for use in preprocessor, and to standardize this logic in one place.

File: test/helpers/lsp.cc [hunk 2]
```diff
@@ -347,4 +347,10 @@ unique_ptr<LSPMessage> makeChange(std::string_view uri, std::string_view content
     return make_unique<LSPMessage>(move(didChangeNotif));
 }
 
+unique_ptr<LSPMessage> makeClose(string_view uri) {
```
Comment (jvilk-stripe) [inline test/helpers/lsp.cc]: Moved from `ProtocolTest.cc` so we can re-use it in different tests.

File: test/lsp/protocol_test_corpus.cc [hunk 3]
```diff
@@ -445,6 +453,40 @@ TEST_F(ProtocolTest, SorbetURIsWork) {
     auto &myMethodDefLoc = myMethodDefinitions.at(0);
     ASSERT_EQ(myMethodDefLoc->uri, "sorbet:folder/foo.rb");
     ASSERT_EQ(readFile(myMethodDefLoc->uri), fileContents);
+
+    // VS Code replaces : in https with something URL-escaped; test that we handle this use-case.
```
Comment (jvilk-stripe) [inline test/lsp/protocol_test_corpus.cc]: While I was at it, I added this test case for this edge case.

File: main/lsp/LSPPreprocessor.cc [hunk 4]
```diff
@@ -356,7 +356,7 @@ void LSPPreprocessor::canonicalizeEdits(u4 v, unique_ptr<DidChangeTextDocumentPa
     updates.versionStart = v;
     updates.versionEnd = v;
     string_view uri = changeParams->textDocument->uri;
-    if (config->isUriInWorkspace(uri)) {
+    if (config->isUriInWorkspace(uri) && !config->isSorbetUri(uri)) {
```
Comment (elliottt) [inline main/lsp/LSPPreprocessor.cc]: This is a little awkward, does `isSorbetUri` need to be included in `isUriInWorkspace`?

Comment (jvilk-stripe) [reply main/lsp/LSPPreprocessor.cc]: Good question. There's one callsite of `isUriInWorkspace` that assume that it accepts `sorbet:` URIs. I think it might be cleaner to change that callsite instead of these ones!

I'll fix that up before submitting.
